cli/src/diff_util.rs

```diff
@@ -1098,23 +1104,40 @@ fn show_unified_diff_hunks(
 pub fn show_git_diff(
     formatter: &mut dyn Formatter,
     store: &Store,
-    tree_diff: TreeDiffStream,
+    from_tree: &MergedTree,
+    to_tree: &MergedTree,
+    matcher: &dyn Matcher,
+    copy_records: &CopyRecords,
     num_context_lines: usize,
 ) -> Result<(), DiffRenderError> {
+    let tree_diff = from_tree.diff_stream(to_tree, matcher, copy_records);
     let mut diff_stream = materialized_diff_stream(store, tree_diff);
+    let copied_sources = collect_copied_sources(copy_records, matcher);
+
     async {
         while let Some(MaterializedTreeDiffEntry {
-            source: _, // TODO handle copy tracking
-            target: path,
+            source: left_path,
+            target: right_path,
             value: diff,
         }) = diff_stream.next().await
         {
-            let path_string = path.as_internal_file_string();
+            let left_path_string = left_path.as_internal_file_string();
+            let right_path_string = right_path.as_internal_file_string();
             let (left_value, right_value) = diff?;
-            let left_part = git_diff_part(&path, left_value)?;
-            let right_part = git_diff_part(&path, right_value)?;
+
+            let left_part = git_diff_part(&left_path, left_value)?;
+            let right_part = git_diff_part(&right_path, right_value)?;
+
+            // Skip the "delete" entry when there is a rename.
+            if right_part.mode.is_none() && copied_sources.contains(left_path.as_ref()) {
+                continue;
+            }
```

Comment:
Slightly off topic, but I think this should be handled by `TreeDiffStream`/`Iterator`. From API standpoint, it's odd that source files are adjusted but the destinations aren't.

(Not meant for this PR)

Reply:
I agree it is odd.  The core issue is that a copy is defined by

source -> still exists
source -> target

and a rename is defined by

source -> removed
source -> target

so the first case doesn't necessarily show up in a diff of the trees at all.

Reply:
I'm not sure if we'll need copy/rename distinction at store layer, but I think we can at least mask "source -> removed" entry by diff stream.

Maybe leave a TODO comment?

Reply:
I think I would prefer to wait for martin and google folks to have there meeting about how to redo the API here, cause the TODO would end up sprinkled in several places.